doc/design/bucketpath.rst

```diff
@@ -0,0 +1,485 @@
+            """
+            Removes this directory.
+
+            Note: In order to stay close to the pathlib based `rmdir` the default won't delete
+                  non empty direcotries, in order to to so recus
+
+            Args:
+                recursive: if true the entrie directory and it's entire contents (files and subdirs)
+                           will be delted. If false and the directory is not empty an error will be thrown.
+
+            Raises:
+                FileNotFoundError: If the file does not exist.
+                PermissionError: If recursive is false and the directory is not empty.
+            """
+
+        def joinpath(*pathsegements) -> Pathlike:
+            """
+            Calling this method is equivalent to combining the path with each of the given pathsegments in turn.
+
+            Returns:
+                A new pathlike object pointing the combined path.
+            """
+
+        def walk() -> Tuple[Pathlike, List[str], List[str]]:
+            """
+            Generate the file names in a directory tree by walking the tree either top-down or bottom-up.
+
+            Note:
+                Try to mimik https://docs.python.org/3/library/pathlib.html#pathlib.Path.walk as closely as possible,
+                except the functionality associated with the parameters of the `pathlib` walk.
+
+            Yields:
+                A 3-tuple of (dirpath, dirnames, filenames).
+            """
+
+        def iterdir() -> Generator[Pathlike, None, None]:
+            """
+            When the path points to a directory, yield path objects of the directory contents.
+
+            Note:
+                For file like paths it will yield nothing.
+
+            Yields:
+                All direct childs of the pathlike object.
+            """
+
+        # Overload / for joining, see also joinpath or `pathlib.Path`.
+        def __truediv__():
+            """
+            """
+
+
+Concrete Paths (Backends)
+-------------------------
+
+Implement the ``Pathlike`` protocol for each specific backend. Concrete paths must include:
+
+- A ``protocol`` member indicating the associated protocol.
+- Backend-specific methods for creation.
+- Implementations for all methods and properties required by the ``Pathlike`` protocol
+
+Each backend must implement the ``as_uri`` method in a way that the location is clearly identifiable.
+
+.. code-block:: python
+
+
+    # Attention: needs to provide/implment Pathlike protocol
+    class BucketPath:
+        """
+        Provides access to a bucket path served via http or https.
+        """
+        
+        # uri protocol specifies associated with this class
+        protocol = ['bfs', 'bfss']
+
+        def __init__(bucket: Bucket, path: str):
+        """
+        Creates a new BucketPath.
+
+        Args:
+            bucket: used for accssing and interacting with to the underlying bucket.
+            path: of the file within the bucket.
+        """
+            
```

Comment:
we might also need a function in PathLike that returns us the path in the UDFs, because often we let the user provide us the information for external access + the bucketfs service name + credentials and use this for first uploading stuff and then get the UDF path to read the uploaded stuff from local file system in the UDF

Reply:
From my pov such a function should not be in `Pathlike`, this would break the abstraction by exposing backend specifics. I would recommend to start out with providing a simple utility function which can do such a thing on the side.

Reply:
All necessary information is contained within the URI, accessible via as_uri(). Therefore, it's feasible to introduce a standalone function, as_local, to handle this. Integrating it into the "interface" later will be straightforward and non-disruptive. Altering the interface signature, in contrast, is of significantly greater impact.

Reply:
extra function which converts http uri to local uri sounds good